fastcrypto/src/bytes_representation.rs

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::encoding::{Base64, Encoding};
+use crate::error::FastCryptoError;
+use crate::serde_helpers::SerializationHelper;
+use serde::de::DeserializeOwned;
+use serde::{de, Deserialize, Serialize};
+use std::fmt::Debug;
+use std::marker::PhantomData;
+
+/// Basic wrapper that stores a bincode serialized version of object T.
+/// To be used in external interfaces instead of the internal object.
+///
+/// Can be derived using [AsBytesRep].
+/// Uses Base64 when serialized with a human readable serializer, and raw bytes otherwise.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct BytesRepresentation<T, const N: usize> {
+    bytes: [u8; N],
+    phantom: PhantomData<T>,
+}
+
+impl<T: Serialize + DeserializeOwned, const N: usize> From<&T> for BytesRepresentation<T, N> {
+    fn from(value: &T) -> Self {
+        // Serialize would fail only if (T, N) is an invalid pair of values, meaning that the type
+        // itself is invalid and therefore the caller has nothing to do with it in runtime.
+        let buffer = bincode::serialize(value).unwrap();
+        Self {
+            bytes: buffer.try_into().unwrap(), // As explained above, this would fail only if (T, N) is an invalid pair of values.
+            phantom: Default::default(),
+        }
+    }
+}
+
+impl<T: Serialize + DeserializeOwned, const N: usize> BytesRepresentation<T, N> {
+    fn bytes_to_type(bytes: &[u8]) -> Result<T, FastCryptoError> {
+        bincode::deserialize(bytes).map_err(|_| FastCryptoError::InvalidInput)
+    }
+
+    pub fn to_type(&self) -> T {
+        // We always check that the byte array represent a valid object before we set it, thus we
+        // will always be able to deserialize it.
+        Self::bytes_to_type(&self.bytes).unwrap()
+    }
+}
+
+impl<T, const N: usize> AsRef<[u8]> for BytesRepresentation<T, N> {
+    fn as_ref(&self) -> &[u8] {
+        self.bytes.as_ref()
+    }
+}
+
+// Define our own serialize/deserialize functions instead of using #[serde_as(as = "Base64")]
+// so we could serialize a flat object (i.e., "1234" instead of "{ bytes: 1234 }").
```

Comment:
i think you can use flatten here? https://serde.rs/attr-flatten.html

Reply:
it's actually an old irrelevant comment, now the serializer does more things.
removed the comments